packages/color-mode/src/color-mode-provider.tsx

```diff
@@ -63,55 +59,48 @@ export function ColorModeProvider(props: ColorModeProviderProps) {
     colorModeManager = localStorageManager,
   } = props
 
+  /**
+   * Only attempt to retrieve if we're on the server. Else this will result
+   * in a hydration mismatch warning and partially invalid visuals.
+   *
+   * Else fallback safely to `theme.config.initialColormode` (default light)
+   */
   const [colorMode, rawSetColorMode] = React.useState<ColorMode | undefined>(
-    () => {
-      /**
-       * Only attempt to retrieve if we're on the server. Else this will result
-       * in a hydration mismatch warning and partially invalid visuals
-       */
-      if (colorModeManager.type === "cookie") {
-        // use `initialColorMode` here to be theme-aware
-        return colorModeManager.get(initialColorMode)
-      }
-
-      if (isBrowser && useSystemColorMode) {
-        return getColorScheme()
-      }
-
-      /**
-       * fallback safely to:
-       * - custom theme.config.initialColorMode
-       * - undefined, will be taken care of in useEffect
-       */
-      return initialColorMode
-    },
+    colorModeManager.type === "cookie"
+      ? colorModeManager.get(initialColorMode)
+      : initialColorMode,
   )
 
   React.useEffect(() => {
     /**
      * Since we cannot initially retrieve localStorage to due above mentioned
      * reasons, do so after hydration.
+     *
+     * Priority:
+     * - system color mode
+     * - defined value on <ColorModeScript />, if present
+     * - previously stored value
      */
-    if (colorModeManager.type === "localStorage") {
-      const root = document.documentElement
-      const mode = root.style.getPropertyValue(
-        "--chakra-ui-color-mode",
-      ) as ColorMode
+    if (isBrowser && colorModeManager.type === "localStorage") {
+      const mode = useSystemColorMode
+        ? getColorScheme()
+        : root.get() || colorModeManager.get()
 
-      rawSetColorMode(mode)
```

Comment:
unconditionally setting the mode here was causing the CSR issues (initially correct, then useEffect sets it to `''` because `.getPropertyValue` for some reason returns `''` if the propety _doesn't exist_